release: Overhaul (#9982)
kubevirt/kubevirt: docs/release.md

> **Note**
> The procedure for cutting a release is automated and can be found [here](release-procedure.md).

# Overview
- [Versioning](#versioning): KubeVirt uses [semantic versioning](http://semver.org)
- [Content](#content): Primary artifact is the source tree in form of signed git tag. Binary artifacts are built for convenience by automation

> - [Platform Support](#platform-support): Every KubeVirt release is targeted towards one Kubernetes release

Reviewer: Shouldn't this mention that two previous Kubernetes versions are also tested?

Reviewer 2: yeah, If we're still planning to test compatiblity with older versions, we should be more explicit about this.
maybe something like... 
`Every KubeVirt release is targeted towards a specific Kubernetes release and is compatible with n-2 kubernetes releases`

Author: Fixed.

These artifacts are provided in their respective channels and their natural way
of distribution.

# Platform Support

> Every KubeVirt release is

Author: Understood, but let me stick to my wording  - no good reasons provided.

 i. targeted towards and life-time coupled with the most recent Kubernetes release
ii. throughout it's life-time the release is validated against the Kubernetes release
    in the [support period](https://kubernetes.io/releases/patch-releases/#support-period)
    at the time of GA of the KubeVirt release

For example:
* KubeVirt v1.0 is targeted towards Kubernetes v1.27.
* KubeVirt v1.0 will be supported until Kubernetes v1.27 runs out of support.
* KubeVirt v1.0 will be validated against all Kubernetes versions being in
  the support period at the time of the v1.0 GA

Leading to:
* At v1.27 GA: v1.0 with v1.27, v1.26, v1.25

> * At v1.28 GA: v1.0 with v1.27, v1.26

Reviewer 3: I wonder if we want to also drop the testing for these EOL versions or do the best effort.

* At v1.30 GA: v1.0 EOL as v1.27 slipped out of the support period.

This is reflected in three areas:

1. _CI_ - A precondition of a release is the presence of CI lanes for the targeted Kubernetes release
2. _Schedule_ - A KubeVirt release will be trailing it's corresponding Kubernetes release. It is trailing it, in order to perform the stabilization on a released Kubernetes version
3. _Support matrix_ - A KubeVirt release's maintenance phase is tied to the maintenance phase (called _support period_) of a Kubernetes release.

The primary reasons for defining a target Kubernetes release are:

- Understanding the target platform in order to perform proper integration work

The primary reasons for defining compatible Kubernetes releases are:

- CI resources are finite. And people to maintain it as well.
- Limit the maintenance burden
- Setting the right expectations with end-users
- Define when a KubeVirt release is ending it's regular support

The Kubernetes release corresponding to a KubeVirt release is defined in the release metadata of a KubeVirt release maintained by [SIG Release](https://github.com/kubevirt/sig-release/).

## Platform Support Skew

> While there are good reasons for defining a target Kubernetes release there are usually also needs to support more than one Kubernetes version.

Reviewer 4: ```suggestion
Although we target a specific Kubernetes release, we recognize that there are reasons to support more than one Kubernetes version.
```

> While the KubeVirt Community can not provide this support, patches to support other releases will also not be rejected as long as they are provided in a reasonable manner (scope and time wise).

Reviewer 4: ```suggestion
While the KubeVirt Community can not provide this support, patches to support other releases will be considered as long as they are provided in a reasonable manner (scope and time wise).
```

> RC and GA tags are create on the stable branch of a release.

Release-Candidate tag examples:

    v0.30.0-rc.0
    v0.30.0-rc.1

GA Release tag example:

    v0.30.0

## Stable branch naming convention

For example, the initial release candidate for branch `release-0.30` is called
`v0.30.1-rc.0`

The determined version is then prefixed with a `v` (mostly for consistency,
because we started this way) and used as the tag name (`$TAG` below).

Branches are created for every minor release and take the form of

    release-<major>.<minor>

For example, the release branch for a v0.30.0 release will be

    release-0.30

Putting all of the above together will lead to the following graph:

```mermaid
gitGraph
  commit
  commit tag: "v0.30-alpha.0"
  commit
  commit tag: "v0.30-alpha.1"
  commit
  commit tag: "v0.30-beta.0"
  commit
  branch release-0.30
  commit tag: "v0.30-rc.0"
  commit
  commit tag: "v0.30-rc.1"
  commit
  commit tag: "v0.30"
  checkout main
  commit
```

# Release Lifecycle

Today a KubeVirt release consists of two phases:

1. _Stabilization phase_ - The timeframe from the stable branch cut all the way towards - and ending with - a new KubeVirt release

> 2. _Maintenance phase_ - The timeframe start with a new KubeVirt release and ending with the release#s End-Of-Life

Reviewer 4: ```suggestion
2. _Maintenance phase_ - The timeframe starts with a new KubeVirt release and ends with the release's [End-Of-Life](#end-of-life-eol)
```

The procedure for handling blockers is described in [release-procedure.md](release-procedure.md).

### Release readiness

The following pre-conditions need to be met before any release tags are created (RC, GA).

#### CI Provider

A pre-condition for the GA of a KubeVirt release is the presence of a KubeVirt CI provider for the corresponding Kubernetes release.
Around the beta release of the corresponding Kubernetes release, the KubeVirt team will start to work on a KubeVirt CI provider.
The assumption is, that the KubeVirt CI provider will be available by KubeVirt's enhancement freeze.

The introduction of a new provider has the following phases:
1. Provider is created
2. Provider is published
3. Provider lands in kubevirt/kubevirt
4. Periodic and presubmit lanes get introduced for the sigs `compute`, `network`, `storage` and `operator`, where
   1. while the periodics deliver a signal for how KubeVirt and the provider are interacting, the presubmits initially are to be triggered manually, so that teams can work on fixing bugs in either the provider or the KubeVirt code
   2. at the point when the periodics look stable enough, the presubmits are turned on to run on every PR

>    3. if the signal is looking stable enough, they are made voting

Reviewer 4: not sure what is meant by "they are made voting"

Author: will gate a PR

Just like in Kubernetes, there will be slowdowns during common holidays which will
cause delays.  So releases that overlap with holidays may be delayed.

### Release procedure

If the release is stable, the release readiness conditions are met, and all holidas are over, then it is finally time to make a release generally available.

The procedure for making a release generally available is described in [release-procedure.md](release-procedure.md).

### Announcement
Every official release must be announced on the `kubevirt-dev` mailinglist
<kubevirt-dev@googlegroups.com> with a body containing the release notes.

You can retrieve the auto generated release notes from the git tag's commit message.

Below is an example of getting the release notes for `v0.31.0-rc.0`

    $ git show v0.31.0-rc.0

## Maintenance phase

The maintenance phase of a KubeVirt release is tied to the [support period](https://kubernetes.io/releases/patch-releases/#support-period) of the corresponding Kubernetes release.

### Bug fixes

During the maintenance phase contributors can backport fixes to a stable branch accoiring to the KubeVirt [backport policy](release-branch-backporting.md).

### Patch releases

> During the maintenance phase KubeVirt will provide patch releases on a irregular basis. 

Reviewer 4: ```suggestion
During the maintenance phase, KubeVirt will provide patch releases on an irregular basis. 
```

### End-Of-Life (EOL)

> A KubeVirt release will reach it's end of life (EOL) once the Kubernetes support period ends.

Reviewer 4: ```suggestion
A KubeVirt release will reach its end of life (EOL) once the Kubernetes support period ends.
```

These artifacts are provided in their respective channels and their natural way
of distribution.

# Platform Support

Every KubeVirt release is targeted towards one specific Kubernetes release.
This is reflected in three areas:

1. _CI_ - A precondition of a release is the presence of CI lanes for the targeted Kubernetes release
2. _Schedule_ - A KubeVirt release will be trailing it's corresponding Kubernetes release. It is trailing it, in order to perform the stabilization on a released Kubernetes version
3. _Support matrix_ - A KubeVirt release's maintenance phase is tied to the maintenance phase of a Kubernetes release.

The primary reasons for defining a target Kubernetes release are:

- CI resources are finite. And people to maintain it as well.
- Limit the maintenance burden
- Understanding the target platform in order to perform proper integration work
- Setting the right expectations with end-users

The Kubernetes release corresponding to a KubeVirt release is defined in the release metadata of a KubeVirt release maintained by [SIG Release](https://github.com/kubevirt/sig-release/).

## Platform Support Skew

> While there are good reasons for defining a target Kubernetes release there are usually also needs to support more than one Kubernetes version.

Reviewer 5: So how will the support be noted in sig-release?  For v1, will it be 1.25, 1.26, 1.27?  1.26, 1.27, 1.28?

I think it would be worth having a link to an example.

Also, we should have a place to note which releases are not supported (if any), eg v1 and 1.27.2

Author: I actually wonder if it's
a)
v1.0 will be targeted at 1.27, and uspported on 1.27, 1.26, 1.25
at 1.28, v1.0 will be supported on 1.27, 1.26. NOT 1.25, as it EOL'ed.
But what about 1.28? Do we backport kubevirt ci to stable branches?

Author: Fixed

Author: It must be a) because the issue of b) is that we get under pressure to potentially backport larger changes in order to keep compatibility with an upcoming release.

> While the KubeVirt Community can not procide this support, patches to support other releases will also not be rejected as long as they are provided in a reasonable manner (scope and time wise).

Reviewer 5: 's/procide/provide/'

Author: Thanks, fixed

> While the KubeVirt Community can not procide this support, patches to support other releases will also not be rejected as long as they are provided in a reasonable manner (scope and time wise).

Reviewer 2: `While the KubeVirt Community can not procide this support` oh interesting. So are we saying we are not supporting N-2 k8s compatiblity?!

Author: We support KubeVirt `N` _on_ Kubernetes `tN`.
And then we support this tuple for the support scope of Kuebrnetes, and thus the support scope of KubeVirt (KubeVirt's scope is tied to Kubernetes support scope).

We would *not* support KubeVirt `N` on Kubenetes`tN-2`

(`N`: A kubeVirt version, `tN` target Kubernetes version of a KubeVirt version)

Reviewer 3: This seems to me like a new policy differing from what we did provide until now. Please correct me if I am misunderstanding this.

Author: I've added an example.
